### ISSUE TITLE
[Bug]: micro stuttering video

### ISSUE THREAD

**Issue by piter-pit** [nova-video-player/aos-AVP#1298]
### Problem description

I have some issues (some people on the internet also) with micro stuttering  when playing h264 or h265 1080p videos, interesting is that it's only on nova player, other players work fine, and I also tried all options like different decoders, frame syncs, buffer sizes etc

### Steps to reproduce the issue

Play any 1080p videos with constant movement

### Your phone/tablet/androidTV model

TCL

### Operating system version

Android 12

### Application version and app store

6.2.97 and 6.3.3 Prerelease

**Comment by reisender1967** [comment]
+1 Same Problem. 
6.3.3 on Amazon Cube

**Comment by courville** [comment]
@reisender1967, is it new with 6.3.3 or there before with official release (latest 6.2.x)?

**Comment by reisender1967** [comment]
It is also with the official release.

I had hoped that the new “Android timely frame” setting would help, but the problem persists.

**Comment by courville** [comment]
@reisender1967, if this is linked to judder, could you try adaptive frame rate options?
And with current state can you please provide your video fps and used screen refresh rate Hz?

**Comment by piter-pit** [comment]
I have just tested it on simple stutter test from YouTube https://youtu.be/FVNO-ehIOdw , downloaded on NAS - same result, copied to TV's internal memory - same result, the bitrate is dramatically low and even then I experience that hiccups. I've recorded in slow motion so if you will not ba able to reproduce or not seeing that please let me know. It's not big stuttering, probably most people would not notice it, my girlfriend is also fine with it 😂, but it can be annoying

**Comment by piter-pit** [comment]
> @reisender1967, if this is linked to judder, could you try adaptive frame rate options? And with current state can you please provide your video fps and used screen refresh rate Hz?

I've tried every option available, even in hidden menu 🫤

**Comment by courville** [comment]
EDIT: @piter-pit and @reisender1967 do you confirm that you have NOT enabled playback speed (I notice a difference with youtube 60Hz stutter video on my phone at 60Hz)?

**Comment by piter-pit** [comment]
> @reisender1967 do you confirm that you have NOT enabled playback speed (I notice a difference with your video on my phone at 60Hz)?

You mean my video or @reisender1967 sent something? In my case playback speed is disabled, was enabled - still stutters, maybe less with disabled but still, similar larger buffer helped a little bit, but I maybe there is different reason

**Comment by courville** [comment]
my bad: @piter-pit youtube 60Hz "stutter" video.

**Comment by piter-pit** [comment]
> my bad: @piter-pit youtube 60Hz "stutter" video.

Ok, and do you see the case? I can try to present it side by side with different player

**Comment by reisender1967** [comment]
> EDIT: @piter-pit and @reisender1967 do you confirm that you have NOT enabled playback speed (I notice a difference with youtube 60Hz stutter video on my phone at 60Hz)?

Playback speed is not enabled. Adaptive frame rate options is on and works. 
So far I have been able to detect the problem with both 24Hz and 50Hz content.

**Comment by courville** [comment]
@piter-pit, if I disable in the "hidden menu" the setting "android timing", I have the 60Hz video stutter test without any jumps on nvidia shield on the tv without adaptive refresh rate. Do you experience the same?
We will have a look at what makes the android timing function nok.

**Comment by piter-pit** [comment]
Hi @courville by hidden menu you mean that additional menu that appears after 6 times clicking on software decoder? I see no option for Android timing unless it's android frame rate feature but then it does not work on my side

**Comment by courville** [comment]
When you click 6 times, you enable advanced options, and there is a new one called Android frame timing at the end of the video section. Disable it and let me know if this is any better please. This option is new to 6.3.3.

**Comment by piter-pit** [comment]
> When you click 6 times, you enable advanced options, and there is a new one called Android frame timing at the end of the video section. Disable it and let me know if this is any better please. This option is new to 6.3.3.

OK, I think I switch that off and on when checking prerelease version also. The effect is the same. Please see video that I recorded side by side on Nova and on a different player. It might not be visible on cell phone screen

https://github.com/user-attachments/assets/3579422a-6158-4032-ae16-e4f63843f374

**Comment by courville** [comment]
@piter-pit thanks for the video and I am also able to reproduce since the beginning.
What I meant is that using the 60fps video on my TV with nvidia shield on 60Hz refresh rate: i) I have the stuttering with "Android frame timing" option on, ii) I do not have the stuttering with "Android frame timing" off. 
This without audio speed option enabled and auto refresh rate disabled.
I would like to know if this is the same on your side.

**Comment by piter-pit** [comment]
It's the same - stuttering with the settings you mentioned.

**Comment by piter-pit** [comment]
BTW there is configuration that gives no stutter android mediaplayer as a decoder. But then I have no sound

**Comment by reisender1967** [comment]
What setting is that? I would like to test it.

**Comment by piter-pit** [comment]
OMG I got it. I mean I went back to all default settings, so: HW decoder, fps matching and android frame timing but disconnected HDMI ARC/eARC cable and no stuttering, when connected with even AVR shutoff - stutters

**Comment by piter-pit** [comment]
> What setting is that? I would like to test it.

In advanced options (to active click 6 times on and off software decoder) then change decoder

**Comment by courville** [comment]
> OMG I got it. I mean I went back to all default settings, so: HW decoder, fps matching and android frame timing but disconnected HDMI ARC/eARC cable and no stuttering, when connected with even AVR shutoff - stutters

that is really strange behavior.
Thus to understand correctly, your setup to watch movies is: TCL android TV running android 12 and HDMI ARC/eARC towards your AV receiver. Can you please disclose your AV receiver model? And please confirm that you are not using any audio passthrough setting (disabled).

**Comment by piter-pit** [comment]
Not sure what is happening but now, after plugged in AVR the stutter test video is still fine but movies are played few fps despite HW decoder... Let me go back to 6.2.97 version

**Comment by courville** [comment]
reboot tv?

**Comment by piter-pit** [comment]
Ok, now when forced passthrough is enabled, I have no few fps, no stutter but also no 5.1, only stereo. It was working fine on AC3 before all this debugging 🫤
When forced disabled, but mode 2 enabled I have few fps

**Comment by courville** [comment]
@piter-pit can you please clarify what you mean by "few fps"? Do you mean that it plays really slow?

**Comment by piter-pit** [comment]
Yes, it's like 2-5 fps frame by frame. I started to use external player but triggered by nova does not load subtitles...
edit: ok, with a help of chatgpt I added script to mix subtitles into video file after nova downloaded it

**Comment by piter-pit** [comment]
On 6.2.90 I was able to set mode 2 passthrough and 5.1 is pass to AVR but stutters are back. Something must be off in audio video sync? But only in nova?

**Comment by reisender1967** [comment]
I can confirm the problem of micro-jerks in connection with the sound.
The juddering is gone when forced passtrough is activated.

**Comment by piter-pit** [comment]
I think micro stutter is over in 6.3.4, however when 5.1 sound should be passthroughed to AVR there is just silence with few fps

**Comment by courville** [comment]
@piter-pit thanks for the feedback.
Could you please describe to me how you connect your TCL TV to your AVR?
I assume it is through eARC with proper HDMI cable.
And could you also tell me what is your TV (not nova) settings in terms of audio output (system audio control setting?) and what are the possible options?
EDIT Cf. https://support.tcl.com/en_US/us-soundbars-troubleshooting/troubleshooting-no-audio-when-connecting-a-tcl-sound-bar-to-a-tv-via-the-arc-or-earc-port https://www.reddit.com/r/Roku/comments/i644n7/tcl_tv_tcl_soundbar_arc_not_passing_audio/

**Comment by piter-pit** [comment]
@courville yes, it's connected via eARC and there are not much to select either passthrough or pcm (stereo). Passthrough is selected. The configuration works fine on any streaming platform (Netflix, Disney, max or prime) and with different player that supports 5.1 output. On nova it also worked 🫤

**Comment by courville** [comment]
@piter-pit, I do not see what I changed for having not working in nova now. It would help if you could pin point the last version that is working.

**Comment by piter-pit** [comment]
@courville as I mentioned few post earlier. In 6.2.90 it was working when set to mode 2

> On 6.2.90 I was able to set mode 2 passthrough and 5.1 is pass to AVR but stutters are back.

**Comment by piter-pit** [comment]
@courville it works fine on 6.3.11, no stuttering as well sound 5.1 is passthrough to AVR so this topic can be closed 🙃
Nevertheless I was wandering if it is complicated to add feature similar like in Vimu where if you AVR does not support certain codec eg my Yamaha does not support multichannel AAC I can set to convert it to EAC-3 5.1 - FFmpeg can do it but not sure how on-the-fly and how complex is integration

**Comment by courville** [comment]
> @courville it works fine on 6.3.11, no stuttering as well sound 5.1 is passthrough to AVR so this topic can be closed 🙃 Nevertheless I was wandering if it is complicated to add feature similar like in Vimu where if you AVR does not support certain codec eg my Yamaha does not support multichannel AAC I can set to convert it to EAC-3 5.1 - FFmpeg can do it but not sure how on-the-fly and how complex is integration

Thanks for the feedback.
AC3 recoding is high on my priority list but not yet addressed...

**Comment by piter-pit** [comment]
AC3 is Dolby digital and should be supported in most AVRs. There might be option to select what kind of codecs you want recode and second menu to select the output codec or maybe even uncompressed pcm.
Let me know if you need any suggestion how setting menu may look like.

**Comment by piter-pit** [comment]
@courville not sure if I didn't notice when quickly checking new version or what, but videos are still stutter 🤔 sorry for mixing up, I think I really had a hope that it is no more stuttering

**Comment by piter-pit** [comment]
@courville Checked version 6.3.13 and the situation is if I select passthrough system encapsulation there is 5.1 but image stutters, if I select nova encapsulation there is no sound (silence) but no stuttering

**Comment by courville** [comment]
@piter-pit, could you please do a test for me? With passthrough disabled, can you check that with 6.3.14, when enabling android frame synchronization in expert options (need to click several times on nova setting software decoding to access export options), on your TCL tv and 60Hz panel refresh rate, can you check if the the playback of 24fps are any better?

**Comment by piter-pit** [comment]
@courville sure. With passthrough disabled there is no stutter, but the sound is stereo only. When passthrough enabled (system encapsulation) it starts stutter. When nova encapsulation sound is just clicking noise

**Comment by piter-pit** [comment]
Just checked old versions: there is same situation in v6.0.62 - no passthrough, no stutter, only stereo and with passthrough, stutters. Checked also lip sync on avr - didn't helped
Edit: on v5.15.21 same situation. Earlier versions I'm not able to install due to android version compatibility

**Comment by piter-pit** [comment]
@courville just did one more check with mi box connected directly to AVR and then AVR to TV, and same result, although a little less visible most likely due to AVR processing, I think. But definitely is.

**Comment by piter-pit** [comment]
@courville just install 6.3.18 and it looks no more stutters. Case can be closed but I'm curious what was the issue since it was for a long time
PS info about AVR capabilities is nice, is it the introduction for audio conversion on the fly in case of not supported format? 😉

**Comment by courville** [comment]
@piter-pit, to solve it we made our avos multimedia engine delegate the frame blitting task to android passing the right muS timing instead of having a lower priority process to manage this (note that it was a little tricky since mkv only output mS frame timing...). Credits to @phhusson who creatively injected the logic in a single place.

**Comment by piter-pit** [comment]
@courville and @phhusson really congrats, I've watched few things and no stutter 🥳 The only thing I've observed is that after pause - the A/V is out of sync until you forward or backward the video, after few pauses it looks like cumulative. Tried different parse and lip sync on/off on AVR. If there is simple fix please consider, otherwise it's a minor issue - I can scroll video after pause 😉

**Comment by phhusson** [comment]
Thanks for the feedback.
I've noticed it as well, and I don't want to roll it out until it's fixed. It's a very off-putting desync because it's the unnatural one, where video is late compared to audio

**Comment by piter-pit** [comment]
@phhusson it's not critical issue at least for me.
BTW, I would like to advocate for a new feature I mentioned earlier in this discussion: adding the ability to convert surround sound on the fly. For example, my AVR doesn't support multichannel AAC, and in Vimu player, there’s an global option to convert such format (or others) to AC3/eAC3 (or other format), or even to uncompressed audio before sending it to AVR. Do you think there’s a chance to add such feature?

**Comment by piter-pit** [comment]
some update after 6.3.20 version: I've noticed that I have to turn on Android frame synchronization (last position in hidden menu, experimental) to have fully smooth playback, with this option disabled there are some issues (still stutters less from time I've created that post) but then, after pause the sound is out of sync.
I think it was ok in 6.3.19 - is it possible?
